jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientResponseListenersTest.java

```diff
@@ -0,0 +1,201 @@
+            .scheme(scenario.getScheme())
+            // Listeners for future events will be notified.
+            .onResponseBegin(r1 -> r1.getRequest().onResponseHeaders(r2 -> beginLatch.countDown()))
+            // Listeners for current or past events won't be notified.
+            .onResponseHeaders(r1 -> r1.getRequest().onResponseHeaders(r2 -> headersLatch.countDown()))
+            .send();
+
+        assertEquals(HttpStatus.OK_200, response.getStatus());
+        assertTrue(beginLatch.await(5, TimeUnit.SECONDS));
+        assertFalse(headersLatch.await(1, TimeUnit.SECONDS));
+    }
+
+    @ParameterizedTest
+    @ArgumentsSource(ScenarioProvider.class)
+    public void testConversationResponse(Scenario scenario) throws Exception
+    {
+        start(scenario, new Handler.Abstract()
+        {
+            @Override
+            public boolean handle(Request request, org.eclipse.jetty.server.Response response, Callback callback) throws Exception
+            {
+                switch (Request.getPathInContext(request))
+                {
+                    case "/old" -> org.eclipse.jetty.server.Response.sendRedirect(request, response, callback, "/new");
+                    case "/new" ->
+                    {
+                        response.getHeaders().put("Special", "Value");
+                        Content.Sink.write(response, true, "data", callback);
+                    }
+                    default -> throw new IllegalStateException();
+                }
+                return true;
+            }
+        });
+
+        ContentResponse response = client.newRequest("localhost", connector.getLocalPort())
+            .scheme(scenario.getScheme())
+            .path("/old")
+            .onResponseBegin(r ->
+            {
+                if (r.getStatus() != HttpStatus.OK_200)
+                    r.abort(new IllegalArgumentException());
+            })
+            .onResponseHeaders(r ->
+            {
+                assertEquals(HttpStatus.OK_200, r.getStatus());
+                if (!r.getHeaders().contains("Special", "Value"))
+                    r.abort(new IllegalArgumentException());
+            })
+            .onResponseContent((r, c) ->
+            {
+                assertEquals(HttpStatus.OK_200, r.getStatus());
+                assertTrue(r.getHeaders().contains("Special", "Value"));
+                if (c == null || !c.hasRemaining())
+                    r.abort(new IllegalArgumentException());
+            })
+            .onResponseSuccess(r ->
+            {
+                assertEquals(HttpStatus.OK_200, r.getStatus());
+                assertTrue(r.getHeaders().contains("Special", "Value"));
+            })
+            .onComplete(result ->
+            {
+                var r = result.getResponse();
+                assertEquals(HttpStatus.OK_200, r.getStatus());
+                assertTrue(r.getHeaders().contains("Special", "Value"));
```

Comment:
It would be good to have an test:
```java
                assertThat(r.getRequest().getPath(), is("/old"));
```

Although, I'm still not sure that the test should not be for "/new"